### PR TITLE
Add Graph Communications SDK infrastructure for Teams meeting audio capture

### LINKED ISSUE #16: Audio Transcription & Meeting Join Features
## Phase 2: Audio Transcription & Meeting Join Features

**Builds on:** #4 (Chat-based Teams bot)

This feature adds real-time audio transcription with speaker diarization to Pennie, enabling her to listen to Teams meeting audio and identify who said what.

---

## Goal

Extend Pennie beyond chat-based interaction to join Teams meetings and transcribe audio in real-time with speaker identification.

**Requirements:**
- Bot joins Teams meetings automatically when invited
- Captures real-time audio streams from the meeting
- Transcribes audio with speaker diarization (identifies who is speaking)
- Streams transcribed text to Pennie agent for processing
- Maintains speaker attribution and timestamps

---

## Prerequisites

Before starting this feature:
- ✅ Issue #4 completed (bot responds to chat messages)
- ✅ Bot deployed and working in Teams
- ✅ Azure AD app registration with Teams bot
- ✅ Backend API integration working

---

### 1. Graph Communications SDK Integration

**Purpose:** Enable bot to join Teams meetings and receive audio streams

**Implementation:**
- Uncomment Graph SDK using statements in `bot/Bots/MediaBot.cs`
- Implement meeting join/leave logic
- Configure media platform for RTP audio stream capture
- Handle meeting invitations
- Manage media stream lifecycle

**Files:**
- `bot/Bots/MediaBot.cs` - Meeting join/leave handlers
- `bot/Services/GraphCallService.cs` - Graph Communications client

**Key Methods:**
```csharp
Task JoinMeetingAsync(string meetingUrl, CancellationToken cancellationToken);
Task LeaveMeetingAsync(CancellationToken cancellationToken);
Task OnAudioReceived(AudioMediaBuffer buffer);
```

**Configuration Required:**
- Media platform settings (RTP ports, audio format)
- Meeting join policies
- Audio stream buffer configuration

---

### 2. Azure Speech Services Integration

**Purpose:** Transcribe audio with speaker diarization

**Implementation:**
- Complete `bot/Services/SpeechTranscriptionService.cs`
- Integrate MeetingTranscriber API
- Configure speaker diarization
- Stream transcription results to Pennie
- Handle audio format conversion (RTP → Speech Services)
- Resolve speaker names from Teams meeting

**Files:**
- `bot/Services/SpeechTranscriptionService.cs` - Speech SDK wrapper
- `bot/Models/TranscriptSegment.cs` - Transcript data model

**Key Methods:**
```csharp
Task StartTranscriptionAsync(CancellationToken cancellationToken);
Task ProcessAudioAsync(byte[] audioData, CancellationToken cancellationToken);
event EventHandler<TranscriptSegment> TranscriptReceived;
```

**Configuration Required:**
- Azure Speech Services key and region
- MeetingTranscriber configuration
- Speaker profile settings
- Audio format specifications (sample rate, channels, encoding)

---

### 3. Audio Pipeline Integration

**Purpose:** Connect audio capture → transcription → Pennie

**Flow:**
```
Teams Meeting Audio
     ↓ RTP Stream (50 frames/sec)
Graph Communications SDK (MediaBot.cs)
     ↓ Audio Data
Azure Speech Services (SpeechTranscriptionService.cs)
     ↓ Transcript + Speaker Names
Pennie AI Agent (via PennieAgentClient.cs)
     ↓ Function Calls (requires_action)
Backend Functions (Azure DevOps)
     ↓ Work Items Created
Teams Chat Notification
```

**Integration Points:**
1. MediaBot receives audio → sends to SpeechTranscriptionService
2. SpeechTranscriptionService transcribes → raises TranscriptReceived event
3. PennieAgentClient sends transcript to Pennie agent
4. Agent processes and calls backend functions
5. Results posted back to Teams chat

---

## Azure AD Permissions

**Required Graph API permissions** (need admin consent):

| Permission | Purpose |
|------------|---------|
| `OnlineMeetings.ReadWrite.All` | Join and manage meetings |
| `Calls.JoinGroupCall.All` | Join group calls |
| `Calls.AccessMedia.All` | Access media streams for audio capture |

**Grant via Azure Portal:**
1. Navigate to Azure AD App Registration
2. API Permissions → Add permission → Microsoft Graph
3. Application permissions → Select permissions above
4. Grant admin consent for tenant

**Verify:**
```bash
az ad app permission list --id {APP_ID}
```

---

### Graph Communications SDK
- [ ] Bot joins Teams meeting when invited
- [ ] Bot leaves meeting gracefully when removed
- [ ] RTP audio streams captured successfully
- [ ] Media platform configured and stable
- [ ] Meeting join/leave logged to Application Insights

### Azure Speech Services
- [ ] Real-time transcription working
- [ ] Speaker diarization identifies participants
- [ ] Speaker names resolved from Teams meeting roster
- [ ] Transcription accuracy acceptable (>90%)
- [ ] Audio format conversion working correctly
- [ ] Transcripts include timestamps

### Integration
- [ ] Audio pipeline flows from meeting → transcription → Pennie
- [ ] PennieAgentClient receives transcript segments
- [ ] Pennie agent processes meeting conversation
- [ ] Function calls triggered based on conversation
- [ ] Work items created with speaker attribution
- [ ] Teams chat notifications posted

### Testing
- [ ] Unit tests for audio processing
- [ ] Integration test: Join meeting → Transcribe → Create work item
- [ ] Load test: Handle 1-hour meeting without crashes
- [ ] Verify speaker attribution in DevOps work items
- [ ] Confirm no audio loss or corruption

### Documentation
- [ ] Document Graph SDK setup and configuration
- [ ] Document Speech Services configuration
- [ ] Update TROUBLESHOOTING.adoc with audio issues
- [ ] Add architecture diagram for audio pipeline

---

## Implementation Steps

1. **Enable Graph SDK** (1 day)
   - Uncomment Graph SDK code in MediaBot.cs
   - Implement meeting join/leave handlers
   - Test meeting join via Bot Framework Emulator
   - Deploy and test in real Teams meeting

2. **Implement Speech Transcription** (1-2 days)
   - Complete SpeechTranscriptionService.cs
   - Integrate MeetingTranscriber API
   - Configure speaker diarization
   - Test audio format conversion
   - Verify transcription quality

3. **Connect Pipeline** (1 day)
   - Wire audio capture → transcription
   - Connect transcription → PennieAgentClient
   - Test end-to-end flow
   - Add error handling and logging

4. **Testing & Refinement** (1 day)
   - Integration tests
   - Load/stress testing
   - Fix any audio quality issues
   - Optimize performance

**Total Estimate:** 4-5 days of focused development

---

## Testing Strategy

1. **Local Testing:**
   - Bot Framework Emulator with mock audio
   - Unit tests for audio processing logic
   - Speech Services API calls with test audio files

2. **Integration Testing:**
   - Create test Teams meeting
   - Invite bot to meeting
   - Speak test phrases and verify transcription
   - Check Pennie receives transcript correctly

3. **End-to-End Testing:**
   - Full meeting scenario with multiple speakers
   - Verify work items created with correct attribution
   - Confirm speaker names match Teams meeting roster
   - Test meeting join/leave edge cases

4. **Performance Testing:**
   - 1-hour meeting without crashes
   - Multiple concurrent meetings (if needed)
   - Memory and CPU usage monitoring

---

### Windows-Only Requirement
- Graph Communications Media SDK only works on Windows
- VM deployment required (already completed in #3)

### Audio Quality
- Network issues can affect audio quality
- Buffer management critical for real-time processing
- Audio format conversion must be precise

### Speaker Diarization Accuracy
- Accuracy depends on audio quality and speaker count
- Background noise can reduce accuracy
- May need fallback for "Unknown Speaker"

### Graph API Rate Limits
- Meeting join operations have rate limits
- Need error handling for throttling
- Consider retry logic with exponential backoff

---

### Microsoft Documentation
- [Graph Communications SDK](https://docs.microsoft.com/en-us/graph/api/resources/communications-api-overview)
- [Azure Speech Services](https://docs.microsoft.com/en-us/azure/cognitive-services/speech-service/)
- [MeetingTranscriber API](https://docs.microsoft.com/en-us/azure/cognitive-services/speech-service/meeting-transcription)
- [Bot Framework Media Bot](https://github.com/microsoftgraph/microsoft-graph-comms-samples)

### Code References
- `bot/Bots/MediaBot.cs` - Main bot logic
- `bot/Services/GraphCallService.cs` - Graph SDK wrapper (placeholder)
- `bot/Services/SpeechTranscriptionService.cs` - Speech SDK wrapper (placeholder)
- `bot/Services/PennieAgentClient.cs` - Agent integration (already working)

### Sample Projects
- [Microsoft Graph Comms Samples](https://github.com/microsoftgraph/microsoft-graph-comms-samples)
- [Teams Media Bot Sample](https://github.com/OfficeDev/Microsoft-Teams-Samples/tree/main/samples/bot-calling-meeting)

---

## Cost Considerations

**Azure Speech Services:**
- ~$1/hour of meeting audio
- Speaker diarization may add additional cost
- Consider using Standard tier for production

**Compute:**
- Windows VM already deployed (#3)
- May need to scale up for multiple concurrent meetings
- Monitor CPU/memory usage during testing

**Estimated Monthly Cost:** $30-150 depending on meeting volume

---

## Success Criteria

This feature is complete when:
- ✅ Bot joins Teams meeting automatically
- ✅ Real-time audio transcription working
- ✅ Speaker diarization identifies participants correctly
- ✅ Transcripts streamed to Pennie agent
- ✅ Work items created with speaker attribution
- ✅ End-to-end testing passes
- ✅ Documentation complete

---

## Related Issues

- **Depends on:** #4 (Chat-based Teams bot)
- **Blocked by:** None (prerequisites met after #4)
- **Related:** #3 (Infrastructure - VM deployed)

---

*This is Phase 2 of the Teams bot implementation. Phase 1 (#4) delivers chat-based interaction, Phase 2 adds audio transcription capabilities.*

### PR DESCRIPTION
**Pull request by BenGWeeks**
## Summary

Implements Issue #16 - Audio Transcription & Meeting Join Features:

- **CallingController** - New controller for Graph Communications SDK notifications
  - `GET /api/calling` - Health check endpoint
  - `POST /api/calling` - Call state notifications from Graph
  - `POST /api/calling/media` - Media notifications endpoint

- **GraphCallService** - Complete rewrite for meeting management
  - MSAL authentication with bot credentials
  - Meeting join URL parsing (extracts ThreadId, TenantId, OrganizerId)
  - Cross-platform compilation with Windows runtime detection
  - Throws `NotImplementedException` on non-Windows (SDK requires Windows Server)

- **MediaBot Integration** - Audio pipeline wiring
  - `TryJoinMeetingForAudioAsync` for automatic meeting detection
  - `OnAudioReceivedAsync` to forward audio to speech transcription
  - `OnTranscriptReceivedAsync` to send results to Pennie agent
  - Proper cleanup on meeting end

- **Configuration** - MediaPlatform settings in appsettings.json
  - ServiceFqdn pointing to Windows VM
  - Call signaling and notification URLs
  - Port configuration for media platform

- **Test Script** - `tests/test-calling-endpoint.sh` for endpoint verification

## Technical Notes

- Graph Communications Media SDK only works on Windows Server
- Code compiles cross-platform but audio capture requires Windows deployment
- SpeechTranscriptionService already implemented with ConversationTranscriber
- Azure AD permissions already configured (Calls.AccessMedia.All, Calls.JoinGroupCall.All, OnlineMeetings.ReadWrite.All)

## Test plan

- [ ] Build succeeds (verified locally)
- [ ] Deploy to Windows VM (pennie-vm-prod)
- [ ] Test `/api/calling` health endpoint returns 200
- [ ] Invite bot to Teams meeting
- [ ] Verify meeting join logs appear
- [ ] Verify audio transcription logs appear (requires Windows)

Closes #16

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: bot/Bots/MediaBot.cs [hunk 1]
```diff
@@ -227,4 +411,381 @@ protected override async Task OnMembersRemovedAsync(
...
+            "can you join"
+        };
+
+        foreach (var pattern in joinPatterns)
+        {
+            if (cleanText.Contains(pattern))
+            {
+                return true;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (BenGWeeks) [reply bot/Bots/MediaBot.cs]: Will leave as-is

File: bot/Services/GraphCallService.cs [hunk 2]
```diff
@@ -87,11 +520,909 @@ public bool IsInMeeting(string meetingId)
...
+                                    "Found organizer ID from calendar event: {OrganizerId}",
+                                    calendarJoinInfo.OrganizerId);
+                                return calendarJoinInfo.OrganizerId;
+                            }
+                            else
+                            {
+                                _logger.LogWarning("Calendar join URL doesn't contain organizer ID (Oid)");
+                            }
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (BenGWeeks) [reply bot/Services/GraphCallService.cs]: Will leave as-is.

File: bot/Services/GraphCallService.cs [hunk 3]
```diff
@@ -87,11 +520,909 @@ public bool IsInMeeting(string meetingId)
...
+                        if (!string.IsNullOrEmpty(stream.SourceId) && uint.TryParse(stream.SourceId, out var msi))
+                        {
+                            if (msiMap.TryAdd(msi, displayName) || msiMap[msi] != displayName)
+                            {
+                                msiMap[msi] = displayName;
+                                _logger.LogInformation("Mapped MSI {MSI} -> {DisplayName} for call {CallId}",
+                                    msi, displayName, callId);
+                            }
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (BenGWeeks) [reply bot/Services/GraphCallService.cs]: Will leave as-is.

File: bot/Services/MediaPlatformService.cs [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+                // Process each speaker's audio separately
+                foreach (var unmixedBuffer in unmixedBuffers)
+                {
+                    if (unmixedBuffer.Length > 0)
+                    {
+                        var audioData = new byte[unmixedBuffer.Length];
+                        Marshal.Copy(unmixedBuffer.Data, audioData, 0, (int)unmixedBuffer.Length);
+                        await callback(audioData);
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (BenGWeeks) [reply bot/Services/MediaPlatformService.cs]: Will leave as-is.
